internal/ingress/controller/parser/kongstate/types.go

```diff
@@ -1,13 +1,45 @@
 package kongstate
 
 import (
+	"fmt"
+
 	"github.com/kong/go-kong/kong"
-	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
+type PortMode int
+
+const (
+	// PortModeImplicit means that the Ingress does not specify the Kubernetes Service port, and that KIC should expect
+	// the Service to have only one port defined.
+	PortModeImplicit PortMode = iota
+	// PortModeNumber means that the Ingress specifies the Service port by raw port number.
+	PortModeByNumber PortMode = iota
+	// PortModeNumber means that the Ingress specifies the Service port by its name field.
+	PortModeByName PortMode = iota
+)
+
+type PortDef struct {
+	Mode PortMode
+
+	// Name is the port name as stated in the Kubernetes service. Must be set iff Mode == PortModeName.
+	Name string
+	// Number is the port number. Must be set iff PortMode == PortModeNumber.
+	Number int32
+}
+
+func (p *PortDef) CanonicalString() string {
+	switch p.Mode {
+	case PortModeByNumber:
+		return fmt.Sprintf("%d", p.Number)
+	case PortModeByName:
+		return p.Name
+	}
+	return "implicitPort"
```

Comment:
This feels like code-smell. Either have an exported constant here or return an empty string.